taglocker_test.go

```diff
@@ -0,0 +1,162 @@
+
+// NoOpReadWriter is a dummy ReadWriter interface that unconditionally succeeds.
+type NoOpReadWriter struct {
+	state uint32
+}
+
+func (norw NoOpReadWriter) ReadTag(name string, value interface{}) error {
+	garbage := norw.state
+	garbage++
+	return nil
+}
+func (norw NoOpReadWriter) WriteTag(name string, value interface{}) error {
+	norw.state++
+	return nil
+}
+
+// Generates a random fully-qualified tag, and returns
+// paths to that tag but also all intermediary prefixes.
+// We will test that we can lock intermediary nodes by
+// randomly choosing from these returned values.
+func genPaths() []string {
+	var paths []string
+
+	/*
+	* The EBNF is:
+	*
+	* tag ::= SYMBOLIC_SEG ( tag_seg )* ( bit_seg )?
+	*
+	* tag_seg ::= '.' SYMBOLIC_SEG
+	*             '[' array_seg ']'
+	*
+	* bit_seg ::= '.' [0-9]+
+	*
+	* array_seg ::= NUMERIC_SEG ( ',' NUMERIC_SEG )*
+	*
+	* SYMBOLIC_SEG ::= [a-zA-Z]([a-zA-Z0-9_]*)
+	*
+	* NUMERIC_SEG ::= [0-9]+
+	*
+	 */
+	genSymbolicSeg := func() string {
+		length := 1 + rand.Intn(10)
+		bs := make([]byte, length)
+		bs[0] = alphaCharset[rand.Intn(len(alphaCharset))]
+		for i := 1; i < len(bs); i++ {
+			bs[i] = alphanumCharset[rand.Intn(len(alphanumCharset))]
+		}
+		return string(bs)
+	}
+
+	genNumericSeg := func() string {
+		return fmt.Sprintf("%d", rand.Intn(10))
+	}
+	genArraySeg := func() string {
+		// TODO: when the parser supports comma-separated multidimensional
+		// arrays, include potentially generating those.
+		return genNumericSeg()
+	}
+	genTagSeg := func() string {
+		if rand.Intn(2) == 0 {
+			return "." + genSymbolicSeg()
+		}
+		return fmt.Sprintf("[%s]", genArraySeg())
+	}
+	// tag ::= SYMBOLIC_SEG ( tag_seg )* ( bit_seg )?
+	var sb strings.Builder
+	sb.WriteString(genSymbolicSeg())
+	for i := 0; i < rand.Intn(5); i++ {
+		//paths = append(paths, sb.String())
+		sb.WriteString(genTagSeg())
+	}
+	paths = append(paths, sb.String())
+	return paths
+}
+
+const serialConcurrency = 1
+const lowConcurrency = 2
+const medConcurrency = 5
+const highConcurrency = 10
+const writePerc = 1
+const heavyWritePerc = 10
+
+func BenchmarkSerialTagLocking(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, serialConcurrency, writePerc)
+}
+
+func BenchmarkSerialTagLockingWithHeavyWrites(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, serialConcurrency, heavyWritePerc)
+}
+
+func BenchmarkLowConcurrencyTagLocking(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, lowConcurrency, writePerc)
+}
+
+func BenchmarkMedConcurrencyTagLocking(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, medConcurrency, writePerc)
+}
+
+func BenchmarkHighConcurrencyTagLocking(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, highConcurrency, writePerc)
+}
+
+func BenchmarkHighConcurrencyTagLockingWithHeavyWrites(b *testing.B) {
+	benchmarkTagLockLocking(b, b.N, highConcurrency, heavyWritePerc)
+}
+
+func benchmarkTagLockLocking(b *testing.B, tests int, concurrency int, writePerc int) {
```

Comment:
I don't think `tests int` should be an argument here. Better to always use `b.N`.